File: MaximumTwinSumOfALinkedList.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def pairSum(self, head: Optional[ListNode]) -> int:

        maxSum = []

        fast = head
        slow = head

        while fast != None:
            fast = fast.next.next
            maxSum.append(slow.val)
            slow = slow.next

        for i in range(len(maxSum)-1 , -1 , -1):
            maxSum[i] += slow.val
            slow = slow.next


        return max(maxSum)

File: MaximumTwinSumOfALinkedList.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    int pairSum(ListNode* head) {
        int max=0;
        vector<int> sum;
        int counter = 0;
        
        while(head != NULL)
        {
            sum.push_back(head->val);
            head = head ->next;
            counter++;
        }
        
        for(int i = 0 ; i < counter/2 ; i++)
        {
            if(sum[i] + sum[counter - i - 1] > max)
                max = sum[i] + sum[counter - i - 1];
        }
        
        return max;
    }
};
